File: python/src/code_smells/dispensables/data_class.py
"""
Data Class Code Smell Example

A class that only contains fields and getters/setters without any behavior.
Other classes manipulate the data class's data directly, violating encapsulation.

Refactoring suggestion: Move Method, Encapsulate Field
"""
from datetime import datetime
from uuid import uuid4


class UserRecord:
    """
    Data class: only stores data, no behavior.

    All logic that operates on this data lives in other classes,
    making the design anemic and non-object-oriented.
    """

    def __init__(self, user_id: str, name: str, email: str, created_at: datetime) -> None:
        self.id = user_id
        self.name = name
        self.email = email
        self.created_at = created_at


class UserService:
    """Service class that manipulates UserRecord data"""

    def create_user(self, name: str, email: str) -> UserRecord:
        if '@' not in email:
            raise ValueError('Invalid email')

        return UserRecord(str(uuid4()), name, email, datetime.now())

    def update_user_email(self, user: UserRecord, new_email: str) -> None:
        if '@' not in new_email:
            raise ValueError('Invalid email')
        user.email = new_email


class UserReportGenerator:
    """Another class manipulating UserRecord data"""

    def generate_user_summary(self, user: UserRecord) -> str:
        return f"User {user.name} ({user.email}) created on {user.created_at.strftime('%Y-%m-%d')}"


def demo_data_class() -> str:
    service = UserService()
    report = UserReportGenerator()
    user = service.create_user('Lina', 'lina@example.com')
    service.update_user_email(user, 'lina+news@example.com')
    return report.generate_user_summary(user)

File: go/code_smells/dispensables/data_class.go
package dispensables

import (
	"fmt"
	"time"

	"github.com/google/uuid"
)

// Code smell: Data Class [Clase de datos].
// UserRecord es una clase que solo almacena datos, mientras que la validación
// y la lógica de creación residen en UserService.

// Ejercicio: Implementa reglas de dominio adicionales, como requerir verificación de email
// o restringir el registro a ciertos dominios (ej. company.com).

// Tendrás que modificar múltiples servicios y lugares que manipulan UserRecord.
// Esto demuestra cómo separar el comportamiento de los datos provoca que cambios simples
// se dispersen ampliamente por el código (Shotgun Surgery).

type UserRecord struct {
	ID        string
	Name      string
	Email     string
	CreatedAt time.Time
}

type UserService struct{}

func (s *UserService) CreateUser(name, email string) (*UserRecord, error) {
	if !contains(email, "@") {
		return nil, fmt.Errorf("Invalid email")
	}

	return &UserRecord{
		ID:        uuid.New().String(),
		Name:      name,
		Email:     email,
		CreatedAt: time.Now(),
	}, nil
}

func (s *UserService) UpdateUserEmail(user *UserRecord, newEmail string) error {
	if !contains(newEmail, "@") {
		return fmt.Errorf("Invalid email")
	}
	user.Email = newEmail
	return nil
}

type UserReportGenerator struct{}

func (g *UserReportGenerator) GenerateUserSummary(user *UserRecord) string {
	return fmt.Sprintf("User %s (%s) created on %s",
		user.Name,
		user.Email,
		user.CreatedAt.Format("2006-01-02"))
}

func DemoDataClass() (string, error) {
	service := &UserService{}
	report := &UserReportGenerator{}
	user, err := service.CreateUser("Lina", "lina@example.com")
	if err != nil {
		return "", err
	}
	err = service.UpdateUserEmail(user, "lina+news@example.com")
	if err != nil {
		return "", err
	}
	return report.GenerateUserSummary(user), nil
}

func contains(s, substr string) bool {
	for i := 0; i <= len(s)-len(substr); i++ {
		if s[i:i+len(substr)] == substr {
			return true
		}
	}
	return false
}
